controllers/api/webhook.go

```diff
@@ -0,0 +1,112 @@
+package api
+
+//TODO
+
+import (
+  "net/http"
+  "strconv"
+  "encoding/json"
+
+  "github.com/gophish/gophish/models"
+  "github.com/gophish/gophish/webhook"
+  "github.com/gorilla/mux"
+  log "github.com/gophish/gophish/logger"
+)
+
+func (as *Server) Webhooks(w http.ResponseWriter, r *http.Request) {
+  switch {
+  case r.Method == "GET":
+    whs, err := models.GetWebhooks()
+    if err != nil {
+      log.Error(err)
+      JSONResponse(w, models.Response{Success: false, Message: err.Error()}, http.StatusInternalServerError)
+      return
+    }
+    JSONResponse(w, whs, http.StatusOK)
+
+  case r.Method == "POST":
+    wh := models.Webhook{}
+    err := json.NewDecoder(r.Body).Decode(&wh)
+    if err != nil {
+      JSONResponse(w, models.Response{Success: false, Message: "Invalid JSON structure"}, http.StatusBadRequest)
+      return
+    }
+    err = models.PostWebhook(&wh)
+    if err != nil {
+      JSONResponse(w, models.Response{Success: false, Message: err.Error()}, http.StatusBadRequest)
+      return
+    }
+    JSONResponse(w, wh, http.StatusCreated)
+  }
+}
+
+func (as *Server) Webhook(w http.ResponseWriter, r *http.Request) {
+  vars := mux.Vars(r)
+  id, _ := strconv.ParseInt(vars["id"], 0, 64)
+  wh, err := models.GetWebhook(id)
+  if err != nil {
+    JSONResponse(w, models.Response{Success: false, Message: "Webhook not found"}, http.StatusNotFound)
+    return
+  }
+  switch {
+  case r.Method == "GET":
+    JSONResponse(w, wh, http.StatusOK)
+
+  case r.Method == "DELETE":
+    err = models.DeleteWebhook(id)
+    if err != nil {
+      JSONResponse(w, models.Response{Success: false, Message: err.Error()}, http.StatusInternalServerError)
+      return
+    }
+    log.Infof("Deleted webhook with id: %d", id)
+    JSONResponse(w, models.Response{Success: true, Message: "Webhook deleted Successfully!"}, http.StatusOK)
+
+  case r.Method == "PUT":
+    wh = models.Webhook{}
+    err = json.NewDecoder(r.Body).Decode(&wh)
+    wh.Id = id;
+    wh.IsActive = false;
+    err = models.PutWebhook(&wh)
+    if err != nil {
+      JSONResponse(w, models.Response{Success: false, Message: err.Error()}, http.StatusBadRequest)
+      return
+    }
+    JSONResponse(w, wh, http.StatusOK)
+  }
+}
+
+func (as *Server) PingWebhook(w http.ResponseWriter, r *http.Request) {
+  switch {
+  case r.Method == "POST":
+    vars := mux.Vars(r)
+    id, _ := strconv.ParseInt(vars["id"], 0, 64)
+    wh, err := models.GetWebhook(id)
+    if err != nil {
+      log.Error(err)
+      JSONResponse(w, models.Response{Success: false, Message: err.Error()}, http.StatusInternalServerError)
+      return
+    }
+
+    httpCnt := &http.Client{} //TODO add timeout and other stuff
```

Comment:
I don't think we should create a new `http.Client` and `webhook.Transport` every time a webhook is used. Instead, here's what I'd recommend:

* We create the `Sender` interface in `webhook/webhook.go` as mentioned in another comment.
* We update the `AdminServer` and `api.Server` to have a `Webhook` field that is a `webhook.Sender`.
* We create a `DefaultSender` in `webhook/webhook.go` that is an `HTTPSender` with a properly configured `http.Client`. You can see an example of this in the [net/http package](https://golang.org/src/net/http/client.go#L107)
* We use this `DefaultSender` when creating the admin server. That way, the same sender is reused for webhook calls.
* We have a `WithWebhook` option for the admin server and api server that sets the `Sender`. This will help us out during testing, since we can use it to set the server's webhook provider to a mock sender.

Reply:
Done. I've not implemented `WithWebhook` yet